v2/coord/behaviour.go

```diff
@@ -0,0 +1,150 @@
+package coord
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+type Notify[C BehaviourEvent] interface {
+	Notify(ctx context.Context, ev C)
+}
+
+type NotifyCloser[C BehaviourEvent] interface {
+	Notify[C]
+	Close()
+}
+
+type NotifyFunc[C BehaviourEvent] func(ctx context.Context, ev C)
+
+func (f NotifyFunc[C]) Notify(ctx context.Context, ev C) {
+	f(ctx, ev)
+}
+
+type Behaviour[I BehaviourEvent, O BehaviourEvent] interface {
+	// Ready returns a channel that signals when the behaviour is ready to perform work.
+	Ready() <-chan struct{}
+
+	// Notify informs the behaviour of an event. The behaviour may perform the event
+	// immediately and queue the result, causing the behaviour to become ready.
+	// It is safe to call Notify from the Perform method.
+	Notify(ctx context.Context, ev I)
+
+	// Perform gives the behaviour the opportunity to perform work or to return a queued
+	// result as an event.
+	Perform(ctx context.Context) (O, bool)
+}
+
+type SM[E any, S any] interface {
+	Advance(context.Context, E) S
+}
+
+type WorkQueueFunc[E BehaviourEvent] func(context.Context, E) bool
+
+// WorkQueue is buffered queue of work to be performed.
+// The queue automatically drains the queue sequentially by calling a
+// WorkQueueFunc for each work item, passing the original context
+// and event.
+type WorkQueue[E BehaviourEvent] struct {
+	pending chan pendingEvent[E]
+	fn      WorkQueueFunc[E]
+	done    atomic.Bool
+	once    sync.Once
+}
+
+func NewWorkQueue[E BehaviourEvent](fn WorkQueueFunc[E]) *WorkQueue[E] {
+	w := &WorkQueue[E]{
+		pending: make(chan pendingEvent[E], 16),
+		fn:      fn,
+	}
+	return w
+}
+
+type pendingEvent[E any] struct {
+	Ctx   context.Context
+	Event E
+}
+
+// Enqueue queues work to be perfomed. It will block if the
+// queue has reached its maximum capacity for pending work. While
+// blocking it will return a context cancellation error if the work
+// item's context is cancelled.
+func (w *WorkQueue[E]) Enqueue(ctx context.Context, cmd E) error {
+	if w.done.Load() {
+		return nil
+	}
+	w.once.Do(func() {
+		go func() {
+			defer w.done.Store(true)
+			for cc := range w.pending {
+				if cc.Ctx.Err() != nil {
+					return
+				}
+				if done := w.fn(cc.Ctx, cc.Event); done {
+					w.done.Store(true)
+					return
```

Comment:
when returning here, anyone sending on `w.pending` will continue to block. The channel needs to be drained here, to release resources of callers of `Enqueue`.

Reply:
Only the waiter sends on the channel since it ownseit and writing is guarded by the done flag. Enqueue exits when done is true. What is the blocking behavior here?

Reply:
Let's imagine I call `Enqueue` a ton of times so that w.pending is full and the `Enqueue `call is blocking in L94. If `fn` returns true for one of the earlier events we return here in L85. The last `Enqueue` calls will continue to block in L94.